File: src/main/cpp/AJAN_Agent/src/AJANParticleUpperBound.cpp
//
// Created by bhuvanesh on 19.06.23.
//

#include "AJANParticleUpperBound.h"

namespace despot {

    AJANParticleUpperBound::AJANParticleUpperBound(const AJANAgent* model):
            agent_model_(model){
        Floor floor = agent_model_->floor_;
        value_.resize(agent_model_->NumStates());
        for(int s=0; s<agent_model_->NumStates(); s++){
            int rob = agent_model_->rob_[s], opp= agent_model_->opp_[s];
            int dist = (int) floor.Distance(rob, opp);
            value_[s] = -(1-Globals::Discount(dist)) /(1-Globals::Discount())
                        + agent_model_->TAG_REWARD * Globals::Discount(dist);
        }
    }

    double AJANParticleUpperBound::Value(const State& s) const {
        const AJANAgentState& state = static_cast<const AJANAgentState&>(s);
        return value_[state.state_id];
    }
}

File: src/main/java/de/dfki/asr/ajan/pluginsystem/mdpplugin/utils/POMDP/implementation/AJANParticleUpperBound.java
package de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.implementation;

import de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.State;
import de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.util.Floor;

import java.util.HashMap;
import java.util.Map;
import java.util.Vector;

import static de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.Problem.Helper.get;
import static java.lang.Math.pow;

public class AJANParticleUpperBound {
    Map<String, Object> variables;
    AJANParticleUpperBound(AJAN_Agent model){
        variables = new HashMap<>();
        Vector<Double> value_ = new Vector<Double>(model.NumStates());
        variables.put("value_",value_);

        Floor floor = (Floor)get("floor_");
        for (int s = 0; s < model.NumStates(); s++) {
            int rob = ((Vector<Integer>)get("rob_")).get(s);
            int opp = ((Vector<Integer>)get("opp_")).get(s);
            int dist = (int) floor.Distance(rob, opp);
            value_.add(s, -(1-discount(dist)/(1-discount())) + (int) get("TAG_REWARD") * discount(dist));
        }
    }

    private double discount(int dist) {
        return pow((int)get("discount"), dist);
    }

    private double discount() {
        return (double) get("discount");
    }

    public double Value(int state_id){
        return ((Vector<Double>)variables.get("value_")).get(state_id);
    }
//    public double Value(AJAN_Agent_State s, int action){
//        return ((Vector<Double>)variables.get("value_")).get(s.state_id);
//    }

}
